NetSSL_OpenSSL/src/HTTPSClientSession.cpp

```diff
@@ -117,6 +117,16 @@ void HTTPSClientSession::abort()
 }
 
 
+StreamSocket HTTPSClientSession::detachSocket()
+{
+	Poco::Net::StreamSocket oldSocket(socket());
+	Poco::Net::Context::Ptr context = static_cast<Poco::Net::SecureStreamSocket>(oldSocket).context();
```

Comment:
This creates a temporary `SecureStreamSocket` from `oldSocket`! Is this really what you want (I guess no, because you wouldn't use `static_cast<>` then).

Reply:
What I really want is to get Context from `oldSocket`, but I can do this only when it's `SecureStreamSocket`.
Casting is not my strong side in C++, how do you suggest it could be done?
What comes to my mind right now is making `SecureStreamSocket oldSocket = static_cast<Poco::Net::SecureStreamSocket>(socket())`, so we can retrive the context and then returning it should work, as it's a subclass of `StreamSocket`, although I'm not 100% sure it's correct.

Reply:
`static_cast<Poco::Net::SecureStreamSocket>(socket())` -- this does not cast the object in the sense that you expect. This is how `static_cast` works: https://en.cppreference.com/w/cpp/language/static_cast . What applies in your case is "1) If there is an implicit conversion sequence from expression to new_type ... static_cast<new_type>(expression) returns the imaginary variable Temp initialized as if by new_type Temp(expression)". It will create a `SecureStreamSocket` and pass `oldSocket` to its constructor.

As you correctly pointed out, "I can do this only when it's SecureStreamSocket". So you should first check whether `socket()` is actually a `SecureStreamSocket()`, and if it is, cast it to a **pointer** or a **reference** to `SecureStreamSocket`. You can do this with `dynamic_cast`:

```c++
auto secure_stream_socket = dynamic_cast<Poco::Net::SecureStreamSocket *>(&socket());
if (secure_stream_socket)
{
   // Ok, now we have a SecureStreamSocket and can call context():
   Poco::Net::Context::Ptr context = secure_stream_socket->context();
}
```